sklearn/preprocessing/tests/test_data.py

```diff
@@ -1948,22 +1949,30 @@ def test_normalizer_max():
 
         for X_norm in (X_norm1, X_norm2):
             row_maxs = X_norm.max(axis=1)
-            for i in range(3):
-                assert_almost_equal(row_maxs[i], 1.0)
-            assert_almost_equal(row_maxs[3], 0.0)
-
-    # check input for which copy=False won't prevent a copy
-    for init in (sparse.coo_matrix, sparse.csc_matrix, sparse.lil_matrix):
-        X = init(X_dense)
-        X_norm = normalizer = Normalizer(norm='l2', copy=False).transform(X)
 
-        assert X_norm is not X
-        assert isinstance(X_norm, sparse.csr_matrix)
-
-        X_norm = toarray(X_norm)
-        for i in range(3):
-            assert_almost_equal(row_maxs[i], 1.0)
-        assert_almost_equal(la.norm(X_norm[3]), 0.0)
+            if X.max() > 0:
+                # skip if the array is all negative
+                for i in range(3):
+                    assert_almost_equal(row_maxs[i], 1.0)
+                assert_almost_equal(row_maxs[3], 0.0)
+
+            # assert the sign does not change after normalization
+            assert_array_almost_equal(
+                np.sign(X_norm1.ravel()), np.sign(toarray(X).ravel()))
+
+        # check input for which copy=False won't prevent a copy
+        for init in (sparse.coo_matrix, sparse.csc_matrix, sparse.lil_matrix):
+            X = init(X_dense)
```

Comment:
So currently this loop is going to happen for `X in (X_dense, X_sparse_pruned` which is redunant. 

Overall this test is getting really complicated. I would rather we reverted it as it was before and added a new simple test based on your example in the issue (and maybe check it for dense and sparse).

Reply:
Alright, I'll revert and add that.